version: "3.8"

services:
  app:
    build: ./app
    container_name: ml_app
    environment:
      - PYTHONPATH=/app
    env_file:
      - .env
    volumes:
      - ./app:/app
      - ./models:/app/models
    ports:
    - "8082:8080"
    depends_on:
      - database
      - rabbitmq
    restart: on-failure

  web-proxy:
    image: nginx:latest
    container_name: nginx_proxy
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf
    ports:
      - "80:80"
      - "443:443"
    depends_on:
      - app
    restart: always

  rabbitmq:
    image: rabbitmq:3-management
    container_name: rabbitmq
    env_file:
      - .env
    volumes:
      - ./data/rabbitmq:/var/lib/rabbitmq
    ports:
      - "15672:15672"
      - "5672:5672"
    restart: on-failure
    healthcheck:
      test: [ "CMD", "rabbitmqctl", "status" ]
      interval: 10s
      timeout: 10s
      retries: 5

  database:
    image: postgres:15
    container_name: postgres_db
    env_file:
      - .env
    volumes:
      - ./data/postgres:/var/lib/postgresql/data
    expose:
      - 5432
    restart: unless-stopped
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -U postgres" ]
      interval: 10s
      timeout: 5s
      retries: 5

  worker:
    build:
      context: ./app
      dockerfile: Dockerfile.worker
    environment:
      - PYTHONPATH=/app
    env_file:
      - .env
    volumes:
      - ./app:/app
      - ./models:/app/models
    depends_on:
      database:
        condition: service_healthy
      rabbitmq:
        condition: service_healthy
    deploy:
      replicas: 2
    restart: on-failure